zerver/lib/subscription_info.py

```diff
@@ -824,19 +809,43 @@ def get_stream_id(sub_dict: RawSubscriptionDict) -> int:
             all_stream_dicts,
             user_profile,
             subscribed_stream_ids,
-            include_subscribers == "partial",
         )
 
+        # Eventually "partial subscribers" will return (at minimum):
+        # (1) all subscriptions for recently active users (for the buddy list)
+        # (2) subscriptions for all bots
+        #
+        # For now, we're only doing (2).
+        send_partial_subscribers = include_subscribers == "partial"
+        partial_subscriber_map: dict[int, list[int]] = dict()
+        if send_partial_subscribers:
+            bot_users = set(
+                UserProfile.objects.filter(
+                    is_bot=True, realm=user_profile.realm, is_active=True
+                ).values_list("id", flat=True)
+            )
+            for stream_id, users in subscriber_map.items():
+                partial_subscribers = [user_id for user_id in users if user_id in bot_users]
+                if len(partial_subscribers) == len(subscriber_map[stream_id]):
+                    partial_subscriber_map[stream_id] = []
+                else:
+                    subscriber_map[stream_id] = []
+                    partial_subscriber_map[stream_id] = partial_subscribers
+
         for lst in [subscribed, unsubscribed]:
             for stream_dict in lst:
                 assert isinstance(stream_dict["stream_id"], int)
                 stream_id = stream_dict["stream_id"]
                 stream_dict["subscribers"] = subscriber_map[stream_id]
+                if send_partial_subscribers:
+                    stream_dict["partial_subscribers"] = partial_subscriber_map[stream_id]
 
         for slim_stream_dict in never_subscribed:
             assert isinstance(slim_stream_dict["stream_id"], int)
             stream_id = slim_stream_dict["stream_id"]
             slim_stream_dict["subscribers"] = subscriber_map[stream_id]
+            if send_partial_subscribers:
+                slim_stream_dict["partial_subscribers"] = partial_subscriber_map[stream_id]
```

Comment:
We will eventually want an `if/else` to not send `subscribers` if we're doing `partial_subscribers`. I don't see much reason to not do that now?

Reply:
I was imagining it as possibly having both -- for each stream we store if there are partial subscribers of if we're sending the full set. We could definitely have situations where some channels are small and only have recently active members and other channels we'd need to do a fetch to get the full subscriber list.

Reply:
Ah yeah, I can do that :)